shared/actions/login/saga.js

```diff
@@ -0,0 +1,588 @@
+    return
+  }
+
+  if (codePage.otherDeviceRole == null) {
+    console.warn("other device role is null, can't setCameraBrokenMode. Bailing")
+    return
+  }
+
+  const mode = defaultModeForDeviceRoles(codePage.myDeviceRole, codePage.otherDeviceRole, broken)
+  if (!mode) {
+    console.warn("mode is null!, can't setCodePageMode. Bailing")
+    return
+  }
+  yield put(Creators.setCodePageMode(mode))
+}
+
+function * loginSuccess () {
+  yield put(Creators.loginDone())
+  yield put(loadDevices())
+  yield put(bootstrap())
+}
+
+function * addNewDeviceSaga ({payload: {role}}: DeviceConstants.AddNewDevice) {
+  yield call(initalizeMyCodeStateForAddingADevice)
+
+  const onBackSaga = function * (response) {
+    yield put(loadDevices())
+    yield put(navigateTo([devicesTab]))
+    if (response) {
+      const engineInst = yield call(engine)
+      yield call([engine, engineInst.cancelRPC], response, InputCancelError)
+    }
+  }
+
+  const finishedSaga = function * () {
+    yield call(onBackSaga)
+  }
+
+  const chooseDeviceTypeSaga = function * ({response}) {
+    const deviceTypeMap: {[key: string]: any} = {
+      [Constants.codePageDeviceRoleNewComputer]: Types.CommonDeviceType.desktop,
+      [Constants.codePageDeviceRoleNewPhone]: Types.CommonDeviceType.mobile,
+    }
+    const deviceType = deviceTypeMap[role]
+
+    yield call(setCodePageOtherDeviceRole, role)
+    yield result(response, deviceType)
+  }
+
+  const addDeviceSagas = {
+    ...kex2Sagas(onBackSaga, onBackSaga, finishedSaga),
+    'keybase.1.provisionUi.chooseDeviceType': chooseDeviceTypeSaga,
+  }
+
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(addDeviceSagas))
+  const addDeviceChanMap = yield call(addDeviceRpc, channelConfig)
+
+  // Returns an Array<Task>
+  yield _.map(addDeviceSagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatestWithCatch(c, finishedSaga, saga), addDeviceChanMap, methodName))
+}
+
+function * reloginSaga ({payload: {usernameOrEmail, passphrase}}: Constants.Relogin) {
+  const finishedSaga = function * ({error}) {
+    if (error) {
+      const message = 'This device is no longer provisioned.'
+      yield put(Creators.loginDone({message}))
+      yield put(Creators.setLoginFromRevokedDevice(message))
+      yield put(navigateTo([loginTab]))
+    } else {
+      yield call(loginSuccess)
+    }
+  }
+
+  const reloginSagas = {
+    'keybase.1.secretUi.getPassphrase': function * ({response}) {
+      yield result(response, {passphrase: passphrase.stringValue(), storeSecret: true})
+    },
+    'finished': finishedSaga,
+  }
+
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(reloginSagas))
+  const chanMap = Types.loginLoginProvisionedDeviceRpcChannelMap(
+    channelConfig,
+    {param: {noPassphrasePrompt: false, username: usernameOrEmail}},
+  )
+
+  yield _.map(reloginSagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatest(c, saga), chanMap, methodName))
```

Comment:
:speech_balloon: A thought: do the channel map / sagas get cleaned up effectively by GC when they are finished?